### DISCUSSION TITLE
umi运行时函数

### DISCUSSION THREAD

**Discussion by lovesL** [umijs/umi#10174] [Q&A]
我很好奇几个运行时的函数是怎么实现 因为我觉得很有用 我经常用到 有没有大佬能简单解答一下 ，或是有没有相关文档贴个地址。 谢谢

**Comment by fz6m** [comment]
你说的 `运行时的函数` 具体是指？

**Reply by lovesL** [reply]
render   patchRoutes  onRouteChange 类似 这些

**Reply by fz6m** [reply]
这些函数所形成的集合可以被称作插件的 runtime 配置部分，`app.ts` 也是一个 runtime 插件。

在 `src/.umi/**` 生成的临时文件中，会有机制去加载这些插件的 runtime 部分然后根据某种策略，比如 `modify` 、`reduce` 瀑布运行，最后的产物用于组装 react 组件，实现原理就是这样的。

**Reply by lovesL** [reply]
虽然听的不是很懂  但是感觉很牛批 。。。大佬 能不能推荐一些文档 我研究研究
